### PR TITLE
refactors ad infinitum

### PR DESCRIPTION
**Pull request by mlugg**
Okay, a bunch of boring changes here, although the last commit ended up being pretty bulky. Perf measurements coming soon (tm). Here's a summary of the changes:

* `AnalSubject` -> `AnalUnit`. I meant to call it this in the first place, I just confused myself. (Any comments bikeshedding this naming will be hidden; I do not care.)
* The frontend now has a better representation of exports. Rather than multiple hashmaps containing pointers, I observed that we only actually need the map formerly known as `export_owners`. By also making the hashmap value an index+length into a larger list rather than an `ArrayListUnmanaged`, we get a much better memory layout which is more friendly to the CPU and better for serialization. Exports are not quite serializable yet, because `Export` itself isn't serializable, but the hard bit is done.
* The above task also led to a surprising amount of backend changes. This includes *minor* fixes in some of the self-hosted linkers, a complete rework of how exports are handled in the C backend to deal with `@export` correctly (thanks @jacobly0!), and a slight refactor of how exports and externs are handled in the LLVM backend to make it a little less broken.
* Move a few datastructures towards `AnalUnit` rather than `DeclIndex`. When we're trying to key something on a "unit of semantic analysis", `AnalUnit` is the correct abstraction, because even though a function has an owner `Decl`, the function has two analyses associated with it: analysis of the function declaration (essentially its signature), and analysis of the function body as a runtime function. Also note that after The Great Decl Split, a generic function instantiation will not have a `Cau` associated with it, since analysis of the instance's signature occurs a; at callsites and b; before the instance even exists! Anyways, the main subject of this change is compile errors -- `failed_decls` becomes `failed_analysis` keyed on `AnalUnit`.
* The reference trace is now tracked very differently. Previously, we had a `reference_table` in `Zcu` which mapped from a `Decl` to the `Decl` that referenced it. Firstly, there's a bit of conflation of `Decl` with `AnalUnit` here, but more importantly, this has a problem when we consider incremental compilation: what if the "referencer" in the table is not actually referenced within the compilation after an update? To give a valid reference trace, we need to be aware of *all* references, so we know all potential "paths" to something being referenced. In fact, storing these gives us a way to *determine* what is referenced at the end of an incremental update.
  
  So, here's how the system works now. `reference_table` maps from the `AnalUnit` **performing** a reference to the `AnalUnit` it references (more precisely, that it triggers semantic analysis of -- immediately or by queuing, it doesn't matter). In fact, it maps to a `u32` index into `all_references`, which contains `Reference` objects, which form a linked list using the `next` field (which is another index into `all_references`). During analysis, when something is analyzed, its entry in `reference_table` is dropped, since these references will be rediscovered (the corresponding indices in `all_references` are put into a freelist). After the main work queue is flushed during an update, `Zcu.resolveReferences` essentially reverses the mapping, returning a hashmap from the `AnalUnit` that **is** referenced to the `AnalUnit` that **performed** the reference. (We don't persist this generally because it's relatively cheap to just construct at the end of compilation and there's a heavy cost associated with maintaining it across incremental updates.) Since we don't have access during this map in the middle of the update (during `Sema.failWithOwnedErrorMsg`), resolution of the reference trace is deferred until `Compilation` gathers errors into the final `ErrorBundle`.
  
  In future, `Zcu.resolveReferences` will do a full traversal, starting from the roots of analysis (the main one being the root file of `std`). That way, we can learn after an incremental update whether any `Decl` is now unreferenced, and if so, skip its stored error messages (and potentially delete it from the binary).
* `Zcu.ErrorMsg` no longer stores a `SrcLoc`, but instead a `LazySrcLoc`. The source location is resolved when the error message is emitted. This is to make it correct across incremental updates.
* `type.zig` becomes `Type.zig`; `@import("type.zig").Type` is now `@import("Type.zig")`. Self-explanatory.
* All types are now queued for full resolution at the moment their definition is analyzed. This simplifies the language specification, particularly in regards to incremental compilation.
* Code generation of functions is now queued into a separate job after the function body is analyzed. This allows us to delete `types_to_resolve` and eliminate all "AIR instruction X triggers type resolution level Y" rules, simplifying the language spec and implementation.
* Type resolution no longer requires a `Sema`, because we construct the `Sema` using the type's owner `Decl`. This is a necessary change with far-reaching consequences in terms of changing function signatures. It fixes bugs exposed by the previous couple of bullet-points, and is necessary for incremental compilation.
* The test case creation logic -- and thus the compiler's build script -- no longer depend on `src/Compilation.zig`. This became a problem because this new logic caused `Compilation` to be fully resolved in the build script, so `@import("build_options")` was analyzed due to its fields, so the build script could not compile. This is fixed by removing this legacy (and quite broken) dependency of the compiler's build script on a random part of its codebase.

### CODE REVIEW

**Comment by mlugg** [comment]
The breaking change here is full type resolution. I previously got the go-ahead from Andrew to make this change without a proposal if it passes the existing tests. It means that, for instance, code like this is now disallowed:
```zig
const S = struct {
    x: 123, // nonsensical type
};
comptime {
    _ = S; // previously didn't resolve S at all; now does!
}
```

**Comment by mlugg** [comment]
## Performance Data Points

### Analyze Behavior
```
Benchmark 1 (51 runs): ../master/stage4/bin/zig test test/behavior.zig --zig-lib-dir lib -fno-emit-bin --test-no-exec
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           395ms ± 12.4ms     384ms …  435ms          4 ( 8%)        0%
  peak_rss            106MB ±  195KB     105MB …  106MB          1 ( 2%)        0%
  cpu_cycles         1.65G  ± 48.9M     1.61G  … 1.81G           5 (10%)        0%
  instructions       2.99G  ± 8.15K     2.99G  … 2.99G           2 ( 4%)        0%
  cache_references    167M  ± 1.60M      165M  …  173M           4 ( 8%)        0%
  cache_misses       7.63M  ±  315K     7.25M  … 8.83M           3 ( 6%)        0%
  branch_misses      7.06M  ± 70.8K     6.93M  … 7.34M           2 ( 4%)        0%
Benchmark 2 (54 runs): ./stage4/bin/zig test test/behavior.zig --zig-lib-dir lib -fno-emit-bin --test-no-exec
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           377ms ± 4.00ms     372ms …  394ms          2 ( 4%)        ⚡-  4.6% ±  0.9%
  peak_rss            109MB ±  225KB     108MB …  109MB          0 ( 0%)        💩+  2.4% ±  0.1%
  cpu_cycles         1.55G  ± 14.9M     1.54G  … 1.62G           3 ( 6%)        ⚡-  5.7% ±  0.8%
  instructions       2.82G  ± 8.33K     2.82G  … 2.82G           0 ( 0%)        ⚡-  5.6% ±  0.0%
  cache_references    162M  ±  885K      160M  …  165M           1 ( 2%)        ⚡-  3.0% ±  0.3%
  cache_misses       7.85M  ±  169K     7.53M  … 8.29M           0 ( 0%)        💩+  2.8% ±  1.3%
  branch_misses      6.77M  ± 42.6K     6.67M  … 6.85M           0 ( 0%)        ⚡-  4.1% ±  0.3%
```

### Compile Behavior (x86_64 selfhosted)
```
Benchmark 1 (23 runs): ../master/stage4/bin/zig test test/behavior.zig --zig-lib-dir lib -fno-llvm -fno-lld --test-no-exec
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           907ms ± 18.6ms     892ms …  976ms          2 ( 9%)        0%
  peak_rss            122MB ±  182KB     121MB …  122MB          0 ( 0%)        0%
  cpu_cycles         3.59G  ± 71.9M     3.55G  … 3.85G           3 (13%)        0%
  instructions       6.56G  ± 49.4K     6.56G  … 6.56G           0 ( 0%)        0%
  cache_references    267M  ± 1.15M      265M  …  269M           0 ( 0%)        0%
  cache_misses       36.5M  ±  569K     35.6M  … 37.9M           0 ( 0%)        0%
  branch_misses      26.7M  ± 97.8K     26.6M  … 26.9M           0 ( 0%)        0%
Benchmark 2 (23 runs): ./stage4/bin/zig test test/behavior.zig --zig-lib-dir lib -fno-llvm -fno-lld --test-no-exec
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           870ms ± 6.99ms     857ms …  888ms          0 ( 0%)        ⚡-  4.1% ±  0.9%
  peak_rss            125MB ±  176KB     125MB …  125MB          0 ( 0%)        💩+  2.8% ±  0.1%
  cpu_cycles         3.43G  ± 25.0M     3.40G  … 3.49G           0 ( 0%)        ⚡-  4.7% ±  0.9%
  instructions       6.62G  ± 49.3K     6.62G  … 6.62G           1 ( 4%)          +  0.9% ±  0.0%
  cache_references    250M  ± 2.05M      248M  …  256M           3 (13%)        ⚡-  6.5% ±  0.4%
  cache_misses       23.7M  ±  280K     23.3M  … 24.4M           1 ( 4%)        ⚡- 35.0% ±  0.7%
  branch_misses      22.6M  ± 96.7K     22.5M  … 22.9M           1 ( 4%)        ⚡- 15.4% ±  0.2%
```

### Analyze Compiler
```
Benchmark 1 (6 runs): /home/mlugg/zig/master/stage4/bin/zig build-exe <snip>
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          5.37s  ± 69.1ms    5.29s  … 5.48s           0 ( 0%)        0%
  peak_rss            237MB ±  244KB     237MB …  237MB          0 ( 0%)        0%
  cpu_cycles         23.7G  ±  491M     23.3G  … 24.6G           0 ( 0%)        0%
  instructions       42.8G  ± 9.71K     42.8G  … 42.8G           0 ( 0%)        0%
  cache_references   2.13G  ± 10.1M     2.11G  … 2.15G           0 ( 0%)        0%
  cache_misses       92.3M  ± 2.94M     88.3M  … 97.1M           0 ( 0%)        0%
  branch_misses      76.8M  ±  319K     76.2M  … 77.1M           0 ( 0%)        0%
Benchmark 2 (6 runs): /home/mlugg/zig/the-great-decl-split/stage4/bin/zig build-exe <snip>
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          5.35s  ± 9.06ms    5.33s  … 5.36s           1 (17%)          -  0.4% ±  1.2%
  peak_rss            274MB ±  242KB     274MB …  274MB          0 ( 0%)        💩+ 15.7% ±  0.1%
  cpu_cycles         23.4G  ± 36.6M     23.3G  … 23.4G           0 ( 0%)          -  1.5% ±  1.9%
  instructions       42.4G  ± 13.1K     42.4G  … 42.4G           0 ( 0%)          -  1.0% ±  0.0%
  cache_references   2.12G  ± 4.85M     2.11G  … 2.12G           0 ( 0%)          -  0.7% ±  0.5%
  cache_misses       98.0M  ± 2.26M     95.2M  …  101M           0 ( 0%)        💩+  6.2% ±  3.7%
  branch_misses      75.1M  ±  445K     74.8M  … 76.0M           0 ( 0%)        ⚡-  2.2% ±  0.6%
```

Overall, it's no worse. Somehow, even though we're surely doing strictly more work, analysis is sometimes faster. (The builds differ only by commits in this branch, and are both identically-created native builds!)

**Comment by mlugg** [comment]
Oh, I did forget to look at the peak RSS in those results. It's a little worse across the board, and noticably worse when rebuilding the compiler.

That's to be expected: we're tracking a lot more state. However, this additional state is necessary for correct incremental compilation, so I don't think this is worth worrying about. Also note that I wasn't using LLVM in those builds, so the absolute peak RSS numbers are relatively low -- that +15% when building the compiler is only actually about 40M more!

**Comment by andrewrk** [comment]
CI failure looks like some linker code got accidentally referenced by an "only_c" build when ensuring that our wasm bootstrap process succeeds. Likely just a missing `if` to turn some branch into a no-op when `build_options.only_c` is true.

File: src/Zcu.zig
Comment (andrewrk) [inline src/Zcu.zig:3699]: Since you have untangled types from Sema instances, maybe it's not so bad to detect this failure in the backends, avoiding this pass entirely. It seems a shame to have a whole pass that accomplishes nothing in the success case.

**Review by andrewrk — APPROVED** [review]
Great work. I have some review feedback, but in the interest of forward progress, I'll proceed with the merge regardless, and you can decide how you want to incorporate the feedback or not in your upcoming changes.

File: src/Zcu.zig
Comment (mlugg) [reply src/Zcu.zig:6496]: In the case where all analysis is succeeded, that's fair -- we could probably avoid this work in that case. However, this is necessary at least in the case where there are entries in `failed_analysis`, because we need to know whether the compile error in question is correct to emit, or whether it's just sitting around from a previous update.

File: src/Sema.zig
Comment (mlugg) [reply src/Sema.zig]: Well, since all types resolve fully anyway, we don't really care which `AnalUnit` caused the *early* resolution, no? We get a valid reference trace regardless.

This change does, I suppose, make error reporting in subtle dependency-loop-esque cases more awkward. That's something we can workshop a solution to when we determine in practice which errors (if any) are not sufficiently detailed.

File: src/Compilation.zig
Comment (mlugg) [reply src/Compilation.zig:324]: It is worth noting that it's important for codegen jobs to go onto the main work queue to make sure type resolution is done before codegen begins. However, once we want to thread codegen, the processing of this job can just grab the `Air` and put it onto the queue for the codegen thread, so the main thread can continue with analysis straight away.

File: src/Zcu.zig
Comment (mlugg) [reply src/Zcu.zig:3699]: Yep, that might be fair. We can investigate that in the future -- indeed, if this branch had a negative performance impact, I was going to investigate that before merge.

File: src/Zcu.zig
Comment (rohlem) [reply src/Zcu.zig:5401]: Afaiu, having an explicit function scope makes it easier for compiler passes to decide whether to inline or deduplicate the code at every callsite/jump to it.
Then is the current behavior of always inlining/duplicating the right default?
Would there be a downside to changing it to automatically pack defers into functions/blocks? (If so, it could also decide with a simple heuristic like statement count, or maybe whether block syntax is used.)

**Comment by andrewrk** [comment]
skipping inclusion in release notes; unclear what should be copy pasted

**Pull request merged**
